Serialization of numpy datetimes

sd2k opened the pull request:
WIP implementation of #182.

Currently just pulls in chrono's serde implementation and uses `Utc.timestamp_nanos`, but I imagine that should be changed. A few things are too naive and don't work right now, such as datetime64's with year, month, and pico/femto/attosecond units.

Feedback is welcome of course, or feel free to close if you're not interested in supporting this and I can maintain a fork.

ijl commented on test/test_numpy.py:
```diff
@@ -117,6 +117,156 @@ def test_numpy_array_d1_bool(self):
+            orjson.dumps(
+                numpy.array(
+                    [
+                        numpy.datetime64("2021"),
```
It would help to generally test minimum year (and I don't know how that compares to datetime.datetime?), 3-digit-or-less (RFC 3339 requires they're zero padded to 4 digits), pre-epoch, something sort of current, and max.

ijl commented on src/serialize/numpy.rs:
```diff
@@ -603,3 +636,115 @@ impl<'p> Serialize for NumpyBool {
+    ///
+    /// It also doesn't work for Picoseconds and smaller right now. Need to
+    /// figure out why.
+    fn num_nanoseconds(&self, val: i64) -> i64 {
```
Looks like a lot of complexity, but as long as the tests are there sure. I don't have the interest or time to fix issues with numpy datetimes later so it needs to really come with extensive tests to be included.

sd2k replied:
The latest commit uses fairly standard chrono functionality to create the datetimes, so if we see e.g. a numpy array with 'year' resolution we just use that year in a call to `chrono::Utc.ymd(<year>, 1, 1)`. Tests should cover several of these cases now.

I think the most common case should remain consistent with the default behaviour for tz-unaware Python datetimes and serialize without a timezone, so I'll switch the logic round to only use Utc when the `OPT_NAIVE_UTC` flag is passed.

ijl commented on src/serialize/numpy.rs:
```diff
@@ -603,3 +636,115 @@ impl<'p> Serialize for NumpyBool {
+        let el0 = ffi!(PyList_GET_ITEM(descr, 0));
+        let descr_str = ffi!(PyTuple_GET_ITEM(el0, 1));
+        let mut str_size: pyo3::ffi::Py_ssize_t = 0;
+        let uni = crate::unicode::read_utf8_from_str(descr_str, &mut str_size);
```
When this stabilizes I suppose you could match this by checking an array of two tuples, that tuple being either the pointer to the PyUnicode or hash of the PyUnicode represented by descr_str, and NumpyDatetimeUnit?

ijl commented on src/serialize/numpy.rs:
```diff
@@ -603,3 +636,115 @@ impl<'p> Serialize for NumpyBool {
+    fn from_pyobject(ptr: *mut PyObject) -> Self {
+        let dtype = ffi!(PyObject_GetAttr(ptr, DTYPE_STR));
+        let descr = ffi!(PyObject_GetAttr(dtype, DESCR_STR));
+        let el0 = ffi!(PyList_GET_ITEM(descr, 0));
```
It's guaranteed the same internal implementation of a datetime within an array yeah?

sd2k replied:
Yeah, numpy chooses the highest resolution unit and stores every element of the array in that unit.

ijl commented on src/serialize/numpy.rs:
```diff
@@ -603,3 +636,115 @@ impl<'p> Serialize for NumpyBool {
+            "ms" => Self::Milliseconds,
+            "us" => Self::Microseconds,
+            "ns" => Self::Nanoseconds,
+            "ps" => Self::Picoseconds,
```
Also if picoseconds is not generally used, fine, it's OK to bail out like with non-C contiguous arrays. The complexity in numpy is enough for compromises in support like this.

sd2k replied:
Cool, that's implemented in the latest commit.

ijl commented:
Looks reasonable to me. Can you support OPT_OMIT_MICROSECONDS as well?

sd2k commented:
I've switched away from using chrono to do the serialization in the latest commit, because it required a heap allocation to do custom formatting (from what I could tell). Instead I've extracted the `write_datetime` functionality from datetime.rs into a `DateTimeLike` trait, which is implemented for both the Python `DateTime` wrapper struct and the `NumpyDateTimeWrapper` struct, so that they both use the existing stack-only serialization implementation.

I wonder if we should introduce an `OPT_OMIT_NANOSECONDS` flag to ignore the nanosecond part of numpy datetimes? Apart from that they're serialized exactly the same as regular Python datetimes now.

ijl commented on Cargo.toml:
```diff
@@ -54,6 +54,7 @@ classifiers = [
 ahash = { version = "0.7", default_features = false }
 associative-cache = { version = "1" }
 bytecount = { version = "^0.6.2", default_features = false, features = ["generic-simd", "runtime-dispatch-simd"] }
+chrono = "0.4.19"
```
Can you specify default_features = false and opt in to what's needed?

ijl reviewed (commented):
Looks good. Needs a rebase and squash and a few comments addressed. I don't see an example that exercises the nanosecond path. I don't know what the use cases there are. It seems simple and reasonable to me to even just drop nanoseconds until someone asks for that resolution rather than add an option there's no clear need for.

sd2k commented:
Thanks for the review. I think I've made all the requested changes, including just dropping nanosecond support (I have no need for it). I did add a comment suggesting how it can be re-added if desired though.

ijl commented:
Thank you. Could you squash to one commit? I'll write documentation after it's merged, but if there's anything you'd like to add please do.

sd2k commented:
Sure, all done!
